pandas/io/stata.py

```diff
@@ -2311,17 +2321,55 @@ def _prepare_pandas(self, data):
             new_type = _convert_datetime_to_stata_type(self._convert_dates[key])
             dtypes[key] = np.dtype(new_type)
 
-        self._set_formats_and_types(data, dtypes)
+        # Verify object arrays are strings and encode to bytes
+        self._encode_strings()
+
+        self._set_formats_and_types(dtypes)
 
         # set the given format for the datetime cols
         if self._convert_dates is not None:
             for key in self._convert_dates:
                 self.fmtlist[key] = self._convert_dates[key]
 
+    def _encode_strings(self):
+        """
+        Encode strings in dta-specific encoding
+
+        Do not encode columns marked for date conversion or for strL
+        conversion. The strL converter independently handles conversion and
+        also accepts empty string arrays.
+        """
+        convert_dates = getattr(self, "_convert_dates", {})
+        convert_strl = getattr(self, "_convert_strl", {})
```

Comment:
it looks like _convert_dates is set on L2102 in `__init__`.  Can we rule out these attributes not existing?

Reply:
Fixed. Thanks.